File: server-csharp/Collisions/GJK Distance/Methods.cs
using System.Diagnostics.Contracts;
using System.Numerics;
using SpacetimeDB;

public static partial class Module
{
    public static bool SolveGjkDistance(List<ConvexHullCollider> ColliderA, DbVector3 PositionA, float YawRadiansA, List<ConvexHullCollider> ColliderB, DbVector3 PositionB, float YawRadiansB, out GjkDistanceResult Result, int MaxIterations = 32)
    {
        float ProgressEpsilon = 1e-4f;

        List<GjkVertex> Simplex = new List<GjkVertex>(4);

        DbVector3 CenterAWorld = GetColliderCenterWorld(new ComplexCollider { ConvexHulls = ColliderA }, PositionA, YawRadiansA);
        DbVector3 CenterBWorld = GetColliderCenterWorld(new ComplexCollider { ConvexHulls = ColliderB }, PositionB, YawRadiansB);

        DbVector3 CenterDelta = Sub(CenterBWorld, CenterAWorld);
        DbVector3 SearchDirection = NearZero(CenterDelta) ? new DbVector3(0f, 1f, 0f) : Normalize(CenterDelta);

        GjkVertex InitialVertex = SupportPairWorld(ColliderA, PositionA, YawRadiansA, ColliderB, PositionB, YawRadiansB, SearchDirection);
        Simplex.Add(InitialVertex);

        for (int IterationIndex = 0; IterationIndex < MaxIterations; IterationIndex++)
        {
            ComputeDistanceFromSimplex(Simplex, out float CurrentDistance, out DbVector3 CurrentPointOnA, out DbVector3 CurrentPointOnB, out DbVector3 CurrentSeparationDirection, out DbVector3 ClosestMinkowskiPoint);

            if (CurrentDistance <= 1e-6f)
            {
                DbVector3 SeparationDirection = NearZero(CurrentSeparationDirection) ? new DbVector3(0f, 1f, 0f) : Normalize(CurrentSeparationDirection);
                Result = new GjkDistanceResult { Intersects = true, Distance = 0f, SeparationDirection = SeparationDirection, PointOnA = CurrentPointOnA, PointOnB = CurrentPointOnB, Simplex = Simplex, LastDirection = SearchDirection };
                return true;
            }

            SearchDirection = Negate(ClosestMinkowskiPoint);
            if (NearZero(SearchDirection))
            {
                Result = new GjkDistanceResult { Intersects = false, Distance = CurrentDistance, SeparationDirection = CurrentSeparationDirection, PointOnA = CurrentPointOnA, PointOnB = CurrentPointOnB, Simplex = Simplex, LastDirection = SearchDirection };
                return true;
            }

            GjkVertex SupportVertex = SupportPairWorld(ColliderA, PositionA, YawRadiansA, ColliderB, PositionB, YawRadiansB, SearchDirection);

            float ClosestDot = Dot(ClosestMinkowskiPoint, SearchDirection);
            float SupportDot = Dot(SupportVertex.MinkowskiPoint, SearchDirection);
            float Progress = SupportDot - ClosestDot;

            if (Progress <= ProgressEpsilon)
            {
                Result = new GjkDistanceResult { Intersects = false, Distance = CurrentDistance, SeparationDirection = CurrentSeparationDirection, PointOnA = CurrentPointOnA, PointOnB = CurrentPointOnB, Simplex = Simplex, LastDirection = SearchDirection };
                return true;
            }

            Simplex.Add(SupportVertex);

            DbVector3 UpdateDirection = SearchDirection;
            if (UpdateSimplex(ref Simplex, ref UpdateDirection))
            {
                DbVector3 SeparationDirection = NearZero(UpdateDirection) ? new DbVector3(0f, 1f, 0f) : Normalize(UpdateDirection);
                Result = new GjkDistanceResult { Intersects = true, Distance = 0f, SeparationDirection = SeparationDirection, PointOnA = SupportVertex.SupportPointA, PointOnB = SupportVertex.SupportPointB, Simplex = Simplex, LastDirection = UpdateDirection };
                return true;
            }
        }

        ComputeDistanceFromSimplex(Simplex, out float FinalDistance, out DbVector3 FinalPointOnA, out DbVector3 FinalPointOnB, out DbVector3 FinalSeparationDirection, out DbVector3 FinalClosestMinkowskiPoint);

        Result = new GjkDistanceResult { Intersects = false, Distance = FinalDistance, SeparationDirection = FinalSeparationDirection, PointOnA = FinalPointOnA, PointOnB = FinalPointOnB, Simplex = Simplex, LastDirection = Negate(FinalClosestMinkowskiPoint) };
        return true;
    }

    static void ComputeDistanceFromSimplex(List<GjkVertex> Simplex, out float Distance, out DbVector3 PointOnA, out DbVector3 PointOnB, out DbVector3 SeparationDirection, out DbVector3 ClosestMinkowskiPoint)
    {
        if (Simplex.Count == 0)
        {
            Distance = 0f;
            PointOnA = new DbVector3(0f, 0f, 0f);
            PointOnB = new DbVector3(0f, 0f, 0f);
            SeparationDirection = new DbVector3(0f, 0f, 1f);
            ClosestMinkowskiPoint = new DbVector3(0f, 0f, 0f);
            return;
        }

        if (Simplex.Count == 1)
        {
            GjkVertex A = Simplex[0];

            ClosestMinkowskiPoint = A.MinkowskiPoint;
            PointOnA = A.SupportPointA;
            PointOnB = A.SupportPointB;

            Distance = Length(ClosestMinkowskiPoint);

            DbVector3 Direction = Negate(ClosestMinkowskiPoint);
            SeparationDirection = NearZero(Direction) ? new DbVector3(0f, 1f, 0f) : Normalize(Direction);
            return;
        }

        if (Simplex.Count == 2)
        {
            ComputeDistanceFromSegment(Simplex[0], Simplex[1], out Distance, out PointOnA, out PointOnB, out SeparationDirection, out ClosestMinkowskiPoint);
            return;
        }

        if (Simplex.Count == 3)
        {
            ComputeDistanceFromTriangle(Simplex[0], Simplex[1], Simplex[2], out Distance, out PointOnA, out PointOnB, out SeparationDirection, out ClosestMinkowskiPoint);
            return;
        }

        ComputeDistanceFromTetrahedron(Simplex[0], Simplex[1], Simplex[2], Simplex[3], out Distance, out PointOnA, out PointOnB, out SeparationDirection, out ClosestMinkowskiPoint);
    }

    static void ComputeDistanceFromSegment(GjkVertex VertexA, GjkVertex VertexB, out float Distance, out DbVector3 PointOnA, out DbVector3 PointOnB, out DbVector3 SeparationDirection, out DbVector3 ClosestMinkowskiPoint)
    {
        DbVector3 A = VertexA.MinkowskiPoint;
        DbVector3 B = VertexB.MinkowskiPoint;
        DbVector3 AB = Sub(B, A);

        float Denominator = Dot(AB, AB);
        float T = 0f;

        if (Denominator > 1e-12f) T = Clamp01(-Dot(A, AB) / Denominator);

        ClosestMinkowskiPoint = Add(A, Mul(AB, T));

        PointOnA = Add(VertexA.SupportPointA, Mul(Sub(VertexB.SupportPointA, VertexA.SupportPointA), T));
        PointOnB = Add(VertexA.SupportPointB, Mul(Sub(VertexB.SupportPointB, VertexA.SupportPointB), T));

        Distance = Length(ClosestMinkowskiPoint);

        DbVector3 Direction = Negate(ClosestMinkowskiPoint);
        SeparationDirection = NearZero(Direction) ? new DbVector3(0f, 1f, 0f) : Normalize(Direction);
    }

    static void ComputeDistanceFromTriangle(GjkVertex VertexA, GjkVertex VertexB, GjkVertex VertexC, out float Distance, out DbVector3 PointOnA, out DbVector3 PointOnB, out DbVector3 SeparationDirection, out DbVector3 ClosestMinkowskiPoint)
    {
        DbVector3 A = VertexA.MinkowskiPoint;
        DbVector3 B = VertexB.MinkowskiPoint;
        DbVector3 C = VertexC.MinkowskiPoint;

        DbVector3 AB = Sub(B, A);
        DbVector3 AC = Sub(C, A);
        DbVector3 AO = Negate(A);

        float D1 = Dot(AB, AO);
        float D2 = Dot(AC, AO);

        if (D1 <= 0f && D2 <= 0f)
        {
            ClosestMinkowskiPoint = A;
            PointOnA = VertexA.SupportPointA;
            PointOnB = VertexA.SupportPointB;
            Distance = Length(ClosestMinkowskiPoint);
            DbVector3 Direction = Negate(ClosestMinkowskiPoint);
            SeparationDirection = NearZero(Direction) ? new DbVector3(0f, 1f, 0f) : Normalize(Direction);
            return;
        }

        DbVector3 BO = Negate(B);
        float D3 = Dot(AB, BO);
        float D4 = Dot(AC, BO);

        if (D3 >= 0f && D4 <= D3)
        {
            ClosestMinkowskiPoint = B;
            PointOnA = VertexB.SupportPointA;
            PointOnB = VertexB.SupportPointB;
            Distance = Length(ClosestMinkowskiPoint);
            DbVector3 Direction = Negate(ClosestMinkowskiPoint);
            SeparationDirection = NearZero(Direction) ? new DbVector3(0f, 1f, 0f) : Normalize(Direction);
            return;
        }

        float VC = D1 * D4 - D3 * D2;
        if (VC <= 0f && D1 >= 0f && D3 <= 0f)
        {
            float V = D1 / (D1 - D3);

            ClosestMinkowskiPoint = Add(A, Mul(AB, V));

            PointOnA = Add(VertexA.SupportPointA, Mul(Sub(VertexB.SupportPointA, VertexA.SupportPointA), V));
            PointOnB = Add(VertexA.SupportPointB, Mul(Sub(VertexB.SupportPointB, VertexA.SupportPointB), V));

            Distance = Length(ClosestMinkowskiPoint);
            DbVector3 Direction = Negate(ClosestMinkowskiPoint);
            SeparationDirection = NearZero(Direction) ? new DbVector3(0f, 1f, 0f) : Normalize(Direction);
            return;
        }

        DbVector3 CO = Negate(C);
        float D5 = Dot(AB, CO);
        float D6 = Dot(AC, CO);

        if (D6 >= 0f && D5 <= D6)
        {
            ClosestMinkowskiPoint = C;
            PointOnA = VertexC.SupportPointA;
            PointOnB = VertexC.SupportPointB;
            Distance = Length(ClosestMinkowskiPoint);
            DbVector3 Direction = Negate(ClosestMinkowskiPoint);
            SeparationDirection = NearZero(Direction) ? new DbVector3(0f, 1f, 0f) : Normalize(Direction);
            return;
        }

        float VB = D5 * D2 - D1 * D6;
        if (VB <= 0f && D2 >= 0f && D6 <= 0f)
        {
            float W = D2 / (D2 - D6);

            ClosestMinkowskiPoint = Add(A, Mul(AC, W));

            PointOnA = Add(VertexA.SupportPointA, Mul(Sub(VertexC.SupportPointA, VertexA.SupportPointA), W));
            PointOnB = Add(VertexA.SupportPointB, Mul(Sub(VertexC.SupportPointB, VertexA.SupportPointB), W));

            Distance = Length(ClosestMinkowskiPoint);
            DbVector3 Direction = Negate(ClosestMinkowskiPoint);
            SeparationDirection = NearZero(Direction) ? new DbVector3(0f, 1f, 0f) : Normalize(Direction);
            return;
        }

        float VA = D3 * D6 - D5 * D4;
        if (VA <= 0f && (D4 - D3) >= 0f && (D5 - D6) >= 0f)
        {
            float W = (D4 - D3) / ((D4 - D3) + (D5 - D6));

            DbVector3 BC = Sub(C, B);
            ClosestMinkowskiPoint = Add(B, Mul(BC, W));

            PointOnA = Add(VertexB.SupportPointA, Mul(Sub(VertexC.SupportPointA, VertexB.SupportPointA), W));
            PointOnB = Add(VertexB.SupportPointB, Mul(Sub(VertexC.SupportPointB, VertexB.SupportPointB), W));

            Distance = Length(ClosestMinkowskiPoint);
            DbVector3 Direction = Negate(ClosestMinkowskiPoint);
            SeparationDirection = NearZero(Direction) ? new DbVector3(0f, 1f, 0f) : Normalize(Direction);
            return;
        }

        float Denominator = VA + VB + VC;
        if (Denominator <= 1e-12f)
        {
            ClosestMinkowskiPoint = A;
            PointOnA = VertexA.SupportPointA;
            PointOnB = VertexA.SupportPointB;
            Distance = Length(ClosestMinkowskiPoint);
            DbVector3 Direction = Negate(ClosestMinkowskiPoint);
            SeparationDirection = NearZero(Direction) ? new DbVector3(0f, 1f, 0f) : Normalize(Direction);
            return;
        }

        float InverseDenominator = 1f / Denominator;
        float VFace = VB * InverseDenominator;
        float WFace = VC * InverseDenominator;
        float UFace = 1f - VFace - WFace;

        ClosestMinkowskiPoint = Add(Add(Mul(A, UFace), Mul(B, VFace)), Mul(C, WFace));

        PointOnA = Add(Add(Mul(VertexA.SupportPointA, UFace), Mul(VertexB.SupportPointA, VFace)), Mul(VertexC.SupportPointA, WFace));
        PointOnB = Add(Add(Mul(VertexA.SupportPointB, UFace), Mul(VertexB.SupportPointB, VFace)), Mul(VertexC.SupportPointB, WFace));

        Distance = Length(ClosestMinkowskiPoint);

        DbVector3 DirectionFace = Negate(ClosestMinkowskiPoint);
        SeparationDirection = NearZero(DirectionFace) ? new DbVector3(0f, 1f, 0f) : Normalize(DirectionFace);
    }

    static void ComputeDistanceFromTetrahedron(GjkVertex VertexA, GjkVertex VertexB, GjkVertex VertexC, GjkVertex VertexD, out float Distance, out DbVector3 PointOnA, out DbVector3 PointOnB, out DbVector3 SeparationDirection, out DbVector3 ClosestMinkowskiPoint)
    {
        float BestDistance = float.MaxValue;

        Distance = 0f;
        PointOnA = new DbVector3(0f, 0f, 0f);
        PointOnB = new DbVector3(0f, 0f, 0f);
        SeparationDirection = new DbVector3(0f, 1f, 0f);
        ClosestMinkowskiPoint = new DbVector3(0f, 0f, 0f);

        EvaluateFace(VertexA, VertexB, VertexC, ref BestDistance, ref Distance, ref PointOnA, ref PointOnB, ref SeparationDirection, ref ClosestMinkowskiPoint);
        EvaluateFace(VertexA, VertexC, VertexD, ref BestDistance, ref Distance, ref PointOnA, ref PointOnB, ref SeparationDirection, ref ClosestMinkowskiPoint);
        EvaluateFace(VertexA, VertexD, VertexB, ref BestDistance, ref Distance, ref PointOnA, ref PointOnB, ref SeparationDirection, ref ClosestMinkowskiPoint);
        EvaluateFace(VertexB, VertexD, VertexC, ref BestDistance, ref Distance, ref PointOnA, ref PointOnB, ref SeparationDirection, ref ClosestMinkowskiPoint);
    }

    static void EvaluateFace(GjkVertex FaceA, GjkVertex FaceB, GjkVertex FaceC, ref float BestDistance, ref float Distance, ref DbVector3 PointOnA, ref DbVector3 PointOnB, ref DbVector3 SeparationDirection, ref DbVector3 ClosestMinkowskiPoint)
    {
        ComputeDistanceFromTriangle(FaceA, FaceB, FaceC, out float FaceDistance, out DbVector3 FacePointOnA, out DbVector3 FacePointOnB, out DbVector3 FaceSeparationDirection, out DbVector3 FaceClosestMinkowskiPoint);

        if (FaceDistance < BestDistance)
        {
            BestDistance = FaceDistance;
            Distance = FaceDistance;
            PointOnA = FacePointOnA;
            PointOnB = FacePointOnB;
            SeparationDirection = FaceSeparationDirection;
            ClosestMinkowskiPoint = FaceClosestMinkowskiPoint;
        }
    }
}

File: server-rust/spacetimedb/src/Collisions/Gjk_Distance/Methods.rs
use crate::*;

pub fn compute_distance_and_unit_separation_from_closest_point(closest_minkowski_point: DbVector3, distance_out: &mut f32, separation_direction_out: &mut DbVector3) {
    let len_sq: f32 = length_sq(closest_minkowski_point);

    if len_sq <= 1e-12 {
        *distance_out = 0.0;
        *separation_direction_out = DbVector3 { x: 0.0, y: 1.0, z: 0.0 };
        return;
    }

    let inv_len: f32 = 1.0 / len_sq.sqrt();
    *distance_out = 1.0 / inv_len;

    let direction = negate(closest_minkowski_point);
    *separation_direction_out = DbVector3 { x: direction.x * inv_len, y: direction.y * inv_len, z: direction.z * inv_len };
}

pub fn solve_gjk_distance(collider_a: &ComplexCollider, position_a: DbVector3, yaw_radians_a: f32, collider_b: &ComplexCollider, position_b: DbVector3, yaw_radians_b: f32, result_out: &mut GjkDistanceResult, max_iterations: i32) -> bool {
    let progress_epsilon: f32 = 1e-4;
    let mut simplex: Vec<GjkVertex> = Vec::with_capacity(4);
    let mut search_direction = DbVector3 { x: 0.0, y: 1.0, z: 0.0 };

    let convex_hulls_a = &collider_a.convex_hulls;
    let convex_hulls_b = &collider_b.convex_hulls;

    simplex.push(support_pair_world(convex_hulls_a, position_a, yaw_radians_a, convex_hulls_b, position_b, yaw_radians_b, search_direction));

    let mut current_distance: f32 = 0.0;
    let mut current_point_on_a = DbVector3 { x: 0.0, y: 0.0, z: 0.0 };
    let mut current_point_on_b = DbVector3 { x: 0.0, y: 0.0, z: 0.0 };
    let mut current_separation_direction = DbVector3 { x: 0.0, y: 1.0, z: 0.0 };
    let mut closest_minkowski_point = DbVector3 { x: 0.0, y: 0.0, z: 0.0 };

    for _iteration_index in 0..max_iterations {
        compute_distance_from_simplex(&simplex, &mut current_distance, &mut current_point_on_a, &mut current_point_on_b, &mut current_separation_direction, &mut closest_minkowski_point);

        if current_distance <= 1e-6 {
            *result_out = GjkDistanceResult { intersects: true, distance: 0.0, separation_direction: current_separation_direction, point_on_a: current_point_on_a, point_on_b: current_point_on_b, simplex: simplex.clone(), last_direction: search_direction };
            return true;
        }

        search_direction = negate(closest_minkowski_point);

        if length_sq(closest_minkowski_point) <= 1e-12 {
            *result_out = GjkDistanceResult { intersects: false, distance: current_distance, separation_direction: current_separation_direction, point_on_a: current_point_on_a, point_on_b: current_point_on_b, simplex: simplex.clone(), last_direction: search_direction };
            return true;
        }

        let support_vertex = support_pair_world(convex_hulls_a, position_a, yaw_radians_a, convex_hulls_b, position_b, yaw_radians_b, search_direction);

        let closest_dot: f32 = dot(closest_minkowski_point, search_direction);
        let support_dot: f32 = dot(support_vertex.minkowski_point, search_direction);

        if support_dot - closest_dot <= progress_epsilon {
            *result_out = GjkDistanceResult { intersects: false, distance: current_distance, separation_direction: current_separation_direction, point_on_a: current_point_on_a, point_on_b: current_point_on_b, simplex: simplex.clone(), last_direction: search_direction };
            return true;
        }

        simplex.push(support_vertex);

        let mut update_direction = search_direction;
        if update_simplex(&mut simplex, &mut update_direction) {
            let separation_direction = normalize_small_vector(update_direction, DbVector3 { x: 0.0, y: 1.0, z: 0.0 });
            let last_vertex = simplex[simplex.len() - 1];
            *result_out = GjkDistanceResult { intersects: true, distance: 0.0, separation_direction, point_on_a: last_vertex.support_point_a, point_on_b: last_vertex.support_point_b, simplex: simplex.clone(), last_direction: update_direction };
            return true;
        }
    }

    compute_distance_from_simplex(&simplex, &mut current_distance, &mut current_point_on_a, &mut current_point_on_b, &mut current_separation_direction, &mut closest_minkowski_point);
    *result_out = GjkDistanceResult { intersects: false, distance: current_distance, separation_direction: current_separation_direction, point_on_a: current_point_on_a, point_on_b: current_point_on_b, simplex, last_direction: negate(closest_minkowski_point) };
    true
}

pub fn compute_distance_from_simplex(simplex: &Vec<GjkVertex>, distance_out: &mut f32, point_on_a_out: &mut DbVector3, point_on_b_out: &mut DbVector3, separation_direction_out: &mut DbVector3, closest_minkowski_point_out: &mut DbVector3) {
    if simplex.len() == 0 {
        *distance_out = 0.0;
        *point_on_a_out = DbVector3 { x: 0.0, y: 0.0, z: 0.0 };
        *point_on_b_out = DbVector3 { x: 0.0, y: 0.0, z: 0.0 };
        *separation_direction_out = DbVector3 { x: 0.0, y: 0.0, z: 1.0 };
        *closest_minkowski_point_out = DbVector3 { x: 0.0, y: 0.0, z: 0.0 };
        return;
    }

    if simplex.len() == 1 {
        let a = simplex[0];

        *closest_minkowski_point_out = a.minkowski_point;
        *point_on_a_out = a.support_point_a;
        *point_on_b_out = a.support_point_b;

        compute_distance_and_unit_separation_from_closest_point(*closest_minkowski_point_out, distance_out, separation_direction_out);
        return;
    }

    if simplex.len() == 2 {
        compute_distance_from_segment(simplex[0], simplex[1], distance_out, point_on_a_out, point_on_b_out, separation_direction_out, closest_minkowski_point_out);
        return;
    }

    if simplex.len() == 3 {
        compute_distance_from_triangle(simplex[0], simplex[1], simplex[2], distance_out, point_on_a_out, point_on_b_out, separation_direction_out, closest_minkowski_point_out);
        return;
    }

    compute_distance_from_tetrahedron(simplex[0], simplex[1], simplex[2], simplex[3], distance_out, point_on_a_out, point_on_b_out, separation_direction_out, closest_minkowski_point_out);
}

pub fn compute_distance_from_segment(vertex_a: GjkVertex, vertex_b: GjkVertex, distance_out: &mut f32, point_on_a_out: &mut DbVector3, point_on_b_out: &mut DbVector3, separation_direction_out: &mut DbVector3, closest_minkowski_point_out: &mut DbVector3) {
    let a = vertex_a.minkowski_point;
    let b = vertex_b.minkowski_point;
    let ab = sub(b, a);

    let denominator: f32 = dot(ab, ab);
    let mut t: f32 = 0.0;

    if denominator > 1e-12 { t = clamp_01(-dot(a, ab) / denominator); }

    *closest_minkowski_point_out = add(a, mul(ab, t));

    *point_on_a_out = add(vertex_a.support_point_a, mul(sub(vertex_b.support_point_a, vertex_a.support_point_a), t));
    *point_on_b_out = add(vertex_a.support_point_b, mul(sub(vertex_b.support_point_b, vertex_a.support_point_b), t));

    compute_distance_and_unit_separation_from_closest_point(*closest_minkowski_point_out, distance_out, separation_direction_out);
}

pub fn compute_distance_from_triangle(vertex_a: GjkVertex, vertex_b: GjkVertex, vertex_c: GjkVertex, distance_out: &mut f32, point_on_a_out: &mut DbVector3, point_on_b_out: &mut DbVector3, separation_direction_out: &mut DbVector3, closest_minkowski_point_out: &mut DbVector3) {
    let a = vertex_a.minkowski_point;
    let b = vertex_b.minkowski_point;
    let c = vertex_c.minkowski_point;

    let ab = sub(b, a);
    let ac = sub(c, a);
    let ao = negate(a);

    let d1: f32 = dot(ab, ao);
    let d2: f32 = dot(ac, ao);

    if d1 <= 0.0 && d2 <= 0.0 {
        *closest_minkowski_point_out = a;
        *point_on_a_out = vertex_a.support_point_a;
        *point_on_b_out = vertex_a.support_point_b;
        compute_distance_and_unit_separation_from_closest_point(*closest_minkowski_point_out, distance_out, separation_direction_out);
        return;
    }

    let bo = negate(b);
    let d3: f32 = dot(ab, bo);
    let d4: f32 = dot(ac, bo);

    if d3 >= 0.0 && d4 <= d3 {
        *closest_minkowski_point_out = b;
        *point_on_a_out = vertex_b.support_point_a;
        *point_on_b_out = vertex_b.support_point_b;
        compute_distance_and_unit_separation_from_closest_point(*closest_minkowski_point_out, distance_out, separation_direction_out);
        return;
    }

    let vc: f32 = d1 * d4 - d3 * d2;
    if vc <= 0.0 && d1 >= 0.0 && d3 <= 0.0 {
        let v: f32 = d1 / (d1 - d3);

        *closest_minkowski_point_out = add(a, mul(ab, v));

        *point_on_a_out = add(vertex_a.support_point_a, mul(sub(vertex_b.support_point_a, vertex_a.support_point_a), v));
        *point_on_b_out = add(vertex_a.support_point_b, mul(sub(vertex_b.support_point_b, vertex_a.support_point_b), v));

        compute_distance_and_unit_separation_from_closest_point(*closest_minkowski_point_out, distance_out, separation_direction_out);
        return;
    }

    let co = negate(c);
    let d5: f32 = dot(ab, co);
    let d6: f32 = dot(ac, co);

    if d6 >= 0.0 && d5 <= d6 {
        *closest_minkowski_point_out = c;
        *point_on_a_out = vertex_c.support_point_a;
        *point_on_b_out = vertex_c.support_point_b;
        compute_distance_and_unit_separation_from_closest_point(*closest_minkowski_point_out, distance_out, separation_direction_out);
        return;
    }

    let vb: f32 = d5 * d2 - d1 * d6;
    if vb <= 0.0 && d2 >= 0.0 && d6 <= 0.0 {
        let w: f32 = d2 / (d2 - d6);

        *closest_minkowski_point_out = add(a, mul(ac, w));

        *point_on_a_out = add(vertex_a.support_point_a, mul(sub(vertex_c.support_point_a, vertex_a.support_point_a), w));
        *point_on_b_out = add(vertex_a.support_point_b, mul(sub(vertex_c.support_point_b, vertex_a.support_point_b), w));

        compute_distance_and_unit_separation_from_closest_point(*closest_minkowski_point_out, distance_out, separation_direction_out);
        return;
    }

    let va: f32 = d3 * d6 - d5 * d4;
    if va <= 0.0 && (d4 - d3) >= 0.0 && (d5 - d6) >= 0.0 {
        let w: f32 = (d4 - d3) / ((d4 - d3) + (d5 - d6));

        let bc = sub(c, b);
        *closest_minkowski_point_out = add(b, mul(bc, w));

        *point_on_a_out = add(vertex_b.support_point_a, mul(sub(vertex_c.support_point_a, vertex_b.support_point_a), w));
        *point_on_b_out = add(vertex_b.support_point_b, mul(sub(vertex_c.support_point_b, vertex_b.support_point_b), w));

        compute_distance_and_unit_separation_from_closest_point(*closest_minkowski_point_out, distance_out, separation_direction_out);
        return;
    }

    let denominator: f32 = va + vb + vc;
    if denominator <= 1e-12 {
        *closest_minkowski_point_out = a;
        *point_on_a_out = vertex_a.support_point_a;
        *point_on_b_out = vertex_a.support_point_b;
        compute_distance_and_unit_separation_from_closest_point(*closest_minkowski_point_out, distance_out, separation_direction_out);
        return;
    }

    let inverse_denominator: f32 = 1.0 / denominator;
    let v_face: f32 = vb * inverse_denominator;
    let w_face: f32 = vc * inverse_denominator;
    let u_face: f32 = 1.0 - v_face - w_face;

    *closest_minkowski_point_out = add(add(mul(a, u_face), mul(b, v_face)), mul(c, w_face));

    *point_on_a_out = add(add(mul(vertex_a.support_point_a, u_face), mul(vertex_b.support_point_a, v_face)), mul(vertex_c.support_point_a, w_face));
    *point_on_b_out = add(add(mul(vertex_a.support_point_b, u_face), mul(vertex_b.support_point_b, v_face)), mul(vertex_c.support_point_b, w_face));

    compute_distance_and_unit_separation_from_closest_point(*closest_minkowski_point_out, distance_out, separation_direction_out);
}

pub fn compute_distance_from_tetrahedron(vertex_a: GjkVertex, vertex_b: GjkVertex, vertex_c: GjkVertex, vertex_d: GjkVertex, distance_out: &mut f32, point_on_a_out: &mut DbVector3, point_on_b_out: &mut DbVector3, separation_direction_out: &mut DbVector3, closest_minkowski_point_out: &mut DbVector3) {
    let mut best_distance: f32 = f32::MAX;

    *distance_out = 0.0;
    *point_on_a_out = DbVector3 { x: 0.0, y: 0.0, z: 0.0 };
    *point_on_b_out = DbVector3 { x: 0.0, y: 0.0, z: 0.0 };
    *separation_direction_out = DbVector3 { x: 0.0, y: 1.0, z: 0.0 };
    *closest_minkowski_point_out = DbVector3 { x: 0.0, y: 0.0, z: 0.0 };

    evaluate_face(vertex_a, vertex_b, vertex_c, &mut best_distance, distance_out, point_on_a_out, point_on_b_out, separation_direction_out, closest_minkowski_point_out);
    evaluate_face(vertex_a, vertex_c, vertex_d, &mut best_distance, distance_out, point_on_a_out, point_on_b_out, separation_direction_out, closest_minkowski_point_out);
    evaluate_face(vertex_a, vertex_d, vertex_b, &mut best_distance, distance_out, point_on_a_out, point_on_b_out, separation_direction_out, closest_minkowski_point_out);
    evaluate_face(vertex_b, vertex_d, vertex_c, &mut best_distance, distance_out, point_on_a_out, point_on_b_out, separation_direction_out, closest_minkowski_point_out);
}

pub fn evaluate_face(face_a: GjkVertex, face_b: GjkVertex, face_c: GjkVertex, best_distance: &mut f32, distance_out: &mut f32, point_on_a_out: &mut DbVector3, point_on_b_out: &mut DbVector3, separation_direction_out: &mut DbVector3, closest_minkowski_point_out: &mut DbVector3) {
    let mut face_distance: f32 = 0.0;
    let mut face_point_on_a = DbVector3 { x: 0.0, y: 0.0, z: 0.0 };
    let mut face_point_on_b = DbVector3 { x: 0.0, y: 0.0, z: 0.0 };
    let mut face_separation_direction = DbVector3 { x: 0.0, y: 1.0, z: 0.0 };
    let mut face_closest_minkowski_point = DbVector3 { x: 0.0, y: 0.0, z: 0.0 };

    compute_distance_from_triangle(face_a, face_b, face_c, &mut face_distance, &mut face_point_on_a, &mut face_point_on_b, &mut face_separation_direction, &mut face_closest_minkowski_point);

    if face_distance < *best_distance {
        *best_distance = face_distance;
        *distance_out = face_distance;
        *point_on_a_out = face_point_on_a;
        *point_on_b_out = face_point_on_b;
        *separation_direction_out = face_separation_direction;
        *closest_minkowski_point_out = face_closest_minkowski_point;
    }
}
